### PR TITLE
[monaco] Re-implement `fireWillSaveModel`

### LINKED ISSUE #3888: [monaco/editor] onWillSave API leads to concurrency problems
VS Code doesn't parallelize edits, and instead they manage a list of listeners/callbacks that get triggered one after the other.

In our case, we first dispatch the event to everyone, then wait for them all. But at this point, every edits are computed at the same time, from the same base document model. This can lead to conflicts, where two contributions want to edit the same region.

I don't really know how to handle that, first thought would be to add an alternative API that would address this issue, and deprecate the other one (but keeping it for some time).

### PR DESCRIPTION
**Pull request by marechal-p**
`fireWillSaveModel` is triggering every listener at the same time, each
providing edits from the same original document, which can lead to
concurrency issues.

Changed the implementation to match what VS Code is doing when
triggering the same listeners: sequentially, and blaming listeners that
directly make modifications to the model.

Signed-off-by: Paul Maréchal <paul.marechal@ericsson.com>

Closes #3888.

### CODE REVIEW

File: packages/monaco/src/browser/monaco-editor-model.ts [hunk 1]
```diff
@@ -52,8 +54,10 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
     protected readonly onDidSaveModelEmitter = new Emitter<monaco.editor.IModel>();
     readonly onDidSaveModel = this.onDidSaveModelEmitter.event;
 
-    protected readonly onWillSaveModelEmitter = new Emitter<WillSaveMonacoModelEvent>();
-    readonly onWillSaveModel = this.onWillSaveModelEmitter.event;
+    protected readonly onWillSaveModelListeners: WillSaveMonacoModelListener[] = [];
+    onWillSaveModel(listener: WillSaveMonacoModelListener) {
```
Comment (akosyakov) [inline packages/monaco/src/browser/monaco-editor-model.ts]: `onWillSaveModel` should have `Event<WillSaveMonacoModelEvent>` type as in vscode API

File: packages/monaco/src/browser/monaco-editor-model.ts [hunk 2]
```diff
@@ -333,28 +337,52 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
...
+                ([] as monaco.editor.IIdentifiedSingleEditOperation[]).concat(...allOperations)
+            );
             if (token.isCancellationRequested) {
                 return;
             }
-            for (const operations of allOperations) {
-                this.applyEdits(operations, {
+
+            // Only apply edits if the document did not change.
+            if (version !== this.version) {
```
Comment (akosyakov) [inline packages/monaco/src/browser/monaco-editor-model.ts]: I don't think we need this check, it should be already ensured by `token.isCancellationRequested`, the operation is cancelled whenever there are changes to the document.

**Comment by marechal-p** [comment]
> What about timing out long-running save participants?

Good idea, in VS Code they have a global timeout, I think in our case it would be the equivalent of triggering the cancellation token.

On the other hand, I was looking at VS Code and the way they handle `Event` objects, turns out the one in Theia is identical, except for one `maxListener` property that feels like a hair in the soup.

Wouldn't it better to extract that property out of the `Event` type and make it specific to what `EventEmitter` objects do?

**Comment by akosyakov** [comment]
> Good idea, in VS Code they have a global timeout, I think in our case it would be the equivalent of triggering the cancellation token.

For the whole save operation or only participants? What happens on timeout the whole operation is cancelled or only participants ignored?

> Wouldn't it better to extract that property out of the Event type and make it specific to what EventEmitter objects do?

`EventEmitter` is an implementation detail, `maxListener` is for clients to configure how many listeners they are expecting to install on a particular event. 

I wonder whether it is still possible to use `EventEmitter` but override `fire` method for example. It does a lot to register/unregister and bind listeners properly already.

**Comment by marechal-p** [comment]
@akosyakov reworked this PR:

I tried to touch as little as possible, can you tell me what you think?

File: packages/monaco/src/browser/monaco-editor-model.ts (hunk 2 above)
Comment (marechal-p) [reply packages/monaco/src/browser/monaco-editor-model.ts]: Apparently I manage to get down to this point, despite the cancellation check before it (I can see an error thrown by this content change check)

**Comment by marechal-p** [comment]
Forgot before, just added a timeout (1 second) for all the listeners to be called.

**Comment by marechal-p** [comment]
Last edit: We want onWillSave listeners to only provide edits, and not change the document. On the other hand, it seems like at least one of our listeners actually does that, changing the document without providing edits.

So I left a soft error displayed in the stderr for now, I don't really know what we should do at that point. Try to better enforce the edit providing, or let it go and not care about listeners making direct modifications?

File: packages/monaco/src/browser/monaco-editor-model.ts (hunk 2 above)
Comment (akosyakov) [reply packages/monaco/src/browser/monaco-editor-model.ts]: ok

File: packages/core/src/common/event.ts [hunk 3]
```diff
@@ -209,7 +222,22 @@ export class Emitter<T> {
...
+    }
+
+    /**
+     * Process each listener one by one.
+     * Return `Emitter.StopSequence` to stop iterating over the listeners.
+     */
+    async sequence(processor: (listener: (e: T) => any) => MaybePromise<any>): Promise<void> {
+        if (this._callbacks) {
+            for (const listener of this._callbacks) {
+                const value = await processor(listener);
```
Comment (akosyakov) [inline packages/core/src/common/event.ts]: Can `processor` return `boolean` instead, when `StopSequence` is redundant?

File: packages/monaco/src/browser/monaco-editor-model.ts [hunk 4]
```diff
@@ -333,28 +333,76 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
...
                 });
             }
         });
+
+        return Promise.race([
+            timeoutPromise,
+            firing,
+        ]).catch(e => {
+            console.error(e);
+        }).then(() => {
```
Comment (akosyakov) [inline packages/monaco/src/browser/monaco-editor-model.ts]: Want the timer cleaned only if an error happened? You can use `await` + `finally`:
```ts
        try {
            await Promise.race([timeoutPromise, firing]);
        } catch (e) {
            console.error(e);
        } finally {
            clearTimeout(timeoutHandle);
        }
```

Comment (marechal-p) [reply packages/monaco/src/browser/monaco-editor-model.ts]: Just tried the following snippet:

```js
Promise.reject(new Error('meh'))
    .catch(e => console.error(e.message))
    .then(() => console.log('done'))

Promise.resolve(1)
    .catch(e => console.error(e.message))
    .then(() => console.log('done'))

// outputs:
meh
done
done
```

The last `then` statement is called every time, acting a bit like a `finally` clause.

But I feel like it looks clearer using `try/catch/finally`, so will do!

**Pull request merged**